src/core/reactor_backend.cc

```diff
@@ -1475,46 +1511,39 @@ class reactor_backend_uring final : public reactor_backend {
     }
     virtual future<size_t> read_some(pollable_fd_state& fd, void* buffer, size_t len) override {
         return readable(fd).then([this, &fd, buffer, len] () mutable {
-            auto r = fd.fd.read(buffer, len);
-            if (!r) {
-                return read_some(fd, buffer, len);
-            }
-            if (size_t(*r) == len) {
+            return queue_request(internal::io_request::make_read(fd.fd.get(), -1, buffer, len, false));
+        }).then([&fd, len] (size_t result) {
+            if (result == len) {
```

Comment:
You can merge the code here into the io_completion subclass. It's already a continuation, no need to add one more here. You'll probably need four subclasses, one for each case.